// projen/lib/web/tailwind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TailwindConfig = void 0;
const json_1 = require("../json");
/**
 * (experimental) Declares a Tailwind CSS configuration file.
 *
 * There are multiple ways to add Tailwind CSS in your node project - see:
 * https://tailwindcss.com/docs/installation
 *
 * @see PostCss
 * @experimental
 */
class TailwindConfig {
    /**
     * @experimental
     */
    constructor(project, options) {
        var _a, _b;
        this.fileName = (_a = options === null || options === void 0 ? void 0 : options.fileName) !== null && _a !== void 0 ? _a : 'tailwind.config.json';
        this.file = new json_1.JsonFile(project, this.fileName, {
            obj: {
                purge: [],
                darkMode: false,
                theme: {
                    extend: {},
                },
                variants: {
                    extend: {},
                },
                plugins: [],
            },
        });
        (_b = project.npmignore) === null || _b === void 0 ? void 0 : _b.exclude(`/${this.fileName}`);
    }
}
exports.TailwindConfig = TailwindConfig;
