// @sls-next/core/dist/route/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.routeDefault = exports.routeApi = exports.handleDomainRedirects = exports.handleAuth = void 0;
const api_1 = require("./api");
const auth_1 = require("./auth");
const basepath_1 = require("./basepath");
const data_1 = require("./data");
const page_1 = require("./page");
const preview_1 = require("./preview");
const redirect_1 = require("./redirect");
const handleAuth = (req, manifest) => {
    const { headers } = req;
    return auth_1.getUnauthenticatedResponse(headers.authorization, manifest.authentication);
};
exports.handleAuth = handleAuth;
const handleCustomRedirects = (req, routesManifest) => {
    const redirect = redirect_1.getRedirectPath(req, routesManifest);
    if (redirect) {
        const { path, statusCode } = redirect;
        return redirect_1.createRedirectResponse(path, req.querystring, statusCode);
    }
};
const handleDomainRedirects = (req, manifest) => {
    const path = redirect_1.getDomainRedirectPath(req, manifest);
    if (path) {
        return redirect_1.createRedirectResponse(path, req.querystring, 308);
    }
};
exports.handleDomainRedirects = handleDomainRedirects;
const handleLanguageRedirect = async (req, manifest, routesManifest) => {
    const languageRedirectUri = await redirect_1.getLanguageRedirectPath(req, manifest, routesManifest);
    if (languageRedirectUri) {
        return redirect_1.createRedirectResponse(languageRedirectUri, req.querystring, 307);
    }
};
const handlePublicFiles = (uri, manifest) => {
    const decodedUri = decodeURI(uri);
    const isPublicFile = manifest.publicFiles && manifest.publicFiles[decodedUri];
    if (isPublicFile) {
        return {
            isPublicFile: true,
            file: uri
        };
    }
};
const handleTrailingSlash = (req, manifest, isFile) => {
    const path = redirect_1.getTrailingSlashPath(req, manifest, isFile);
    if (path) {
        return redirect_1.createRedirectResponse(path, req.querystring, 308);
    }
};
/*
 * Routes:
 * - auth
 * - redirects
 * - api routes
 * - rewrites (external and api)
 */
const routeApi = (req, manifest, routesManifest) => {
    const auth = exports.handleAuth(req, manifest);
    if (auth) {
        return auth;
    }
    const redirect = exports.handleDomainRedirects(req, manifest) ||
        handleCustomRedirects(req, routesManifest);
    if (redirect) {
        return redirect;
    }
    return api_1.handleApiReq(req.uri, manifest, routesManifest);
};
exports.routeApi = routeApi;
/*
 * Routes:
 * - auth
 * - redirects
 * - public files
 * - data routes
 * - pages
 * - rewrites (external and page)
 */
const routeDefault = async (req, manifest, prerenderManifest, routesManifest) => {
    const auth = exports.handleAuth(req, manifest);
    if (auth) {
        return auth;
    }
    const domainRedirect = exports.handleDomainRedirects(req, manifest);
    if (domainRedirect) {
        return domainRedirect;
    }
    const uri = basepath_1.normalise(req.uri, routesManifest);
    const is404 = uri.endsWith("/404");
    const isDataReq = uri.startsWith("/_next/data");
    const publicFile = handlePublicFiles(uri, manifest);
    const isPublicFile = !!publicFile;
    const trailingSlash = !is404 && handleTrailingSlash(req, manifest, isDataReq || isPublicFile);
    if (trailingSlash) {
        return trailingSlash;
    }
    if (publicFile) {
        return publicFile;
    }
    const otherRedirect = handleCustomRedirects(req, routesManifest) ||
        (await handleLanguageRedirect(req, manifest, routesManifest));
    if (otherRedirect) {
        return otherRedirect;
    }
    const isPreview = await preview_1.isValidPreviewRequest(req.headers.cookie, prerenderManifest.preview.previewModeSigningKey);
    if (isDataReq) {
        return data_1.handleDataReq(uri, manifest, routesManifest, isPreview);
    }
    else {
        return page_1.handlePageReq(req.uri, manifest, routesManifest, isPreview);
    }
};
exports.routeDefault = routeDefault;
