// projen/lib/jest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Jest = void 0;
const path = require("path");
const tasks_1 = require("./tasks");
const typescript_1 = require("./typescript");
const DEFAULT_TEST_REPORTS_DIR = 'test-reports';
/**
 * (experimental) Installs the following npm scripts:.
 *
 * - `test` will run `jest --passWithNoTests`
 * - `test:watch` will run `jest --watch`
 * - `test:update` will run `jest -u`
 *
 * @experimental
 */
class Jest {
    /**
     * @experimental
     */
    constructor(project, options = {}) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        this.project = project;
        const jestDep = options.jestVersion ? `jest@${options.jestVersion}` : 'jest';
        project.addDevDeps(jestDep);
        this.ignorePatterns = (_a = options.ignorePatterns) !== null && _a !== void 0 ? _a : ['/node_modules/'];
        this.jestConfig = options.jestConfig;
        this.typescriptConfig = options.typescriptConfig;
        const coverageDirectory = (_c = (_b = this.jestConfig) === null || _b === void 0 ? void 0 : _b.coverageDirectory) !== null && _c !== void 0 ? _c : 'coverage';
        this.reporters = [];
        if ((_d = options.preserveDefaultReporters) !== null && _d !== void 0 ? _d : true) {
            this.reporters.unshift('default');
        }
        this.config = {
            ...this.jestConfig,
            clearMocks: (_f = (_e = this.jestConfig) === null || _e === void 0 ? void 0 : _e.clearMocks) !== null && _f !== void 0 ? _f : true,
            collectCoverage: (_j = (_g = options.coverage) !== null && _g !== void 0 ? _g : (_h = this.jestConfig) === null || _h === void 0 ? void 0 : _h.collectCoverage) !== null && _j !== void 0 ? _j : true,
            coverageDirectory: coverageDirectory,
            coveragePathIgnorePatterns: (_l = (_k = this.jestConfig) === null || _k === void 0 ? void 0 : _k.coveragePathIgnorePatterns) !== null && _l !== void 0 ? _l : this.ignorePatterns,
            testPathIgnorePatterns: (_o = (_m = this.jestConfig) === null || _m === void 0 ? void 0 : _m.testPathIgnorePatterns) !== null && _o !== void 0 ? _o : this.ignorePatterns,
            testMatch: (_q = (_p = this.jestConfig) === null || _p === void 0 ? void 0 : _p.testMatch) !== null && _q !== void 0 ? _q : [
                '**/__tests__/**/*.js?(x)',
                '**/?(*.)+(spec|test).js?(x)',
            ],
            reporters: this.reporters,
        };
        if ((_r = options.junitReporting) !== null && _r !== void 0 ? _r : true) {
            const reportsDir = DEFAULT_TEST_REPORTS_DIR;
            this.addReporter([
                'jest-junit',
                { outputDirectory: reportsDir },
            ]);
            project.addDevDeps('jest-junit@^12');
            project.gitignore.exclude('# jest-junit artifacts', `/${reportsDir}/`, 'junit.xml');
        }
        if ((_s = this.jestConfig) === null || _s === void 0 ? void 0 : _s.reporters) {
            for (const reporter of this.jestConfig.reporters) {
                this.addReporter(reporter);
            }
        }
        if ((_t = this.jestConfig) === null || _t === void 0 ? void 0 : _t.coverageThreshold) {
            this.config.coverageThreshold = {
                global: (_u = this.jestConfig) === null || _u === void 0 ? void 0 : _u.coverageThreshold,
            };
        }
        this.configureTestCommand();
        project.addFields({ jest: this.config });
        const coverageDirectoryPath = path.join('/', coverageDirectory);
        (_v = project.npmignore) === null || _v === void 0 ? void 0 : _v.exclude(coverageDirectoryPath);
        project.gitignore.exclude(coverageDirectoryPath);
        project.addTip('The VSCode jest extension watches in the background and shows inline test results');
    }
    /**
     * @experimental
     */
    addIgnorePattern(pattern) {
        this.ignorePatterns.push(pattern);
    }
    /**
     * @experimental
     */
    addReporter(reporter) {
        this.reporters.push(reporter);
    }
    /**
     * (experimental) Merges passed in typescript config options with jest configured typescript options from .projenrc Add Jest config settings for typescript options.
     *
     * @param options TypescriptConfigOptions.
     * @experimental
     */
    generateTypescriptConfig(options) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const tsconfig = new typescript_1.TypescriptConfig(this.project, {
            fileName: (_a = options.fileName) !== null && _a !== void 0 ? _a : 'tsconfig.jest.json',
            include: [
                ...options.include ? options.include : [],
                ...((_b = this.typescriptConfig) === null || _b === void 0 ? void 0 : _b.include) ? (_c = this.typescriptConfig) === null || _c === void 0 ? void 0 : _c.include : [],
            ],
            exclude: [
                ...options.exclude ? options.exclude : [],
                ...((_d = this.typescriptConfig) === null || _d === void 0 ? void 0 : _d.exclude) ? (_e = this.typescriptConfig) === null || _e === void 0 ? void 0 : _e.exclude : [],
            ],
            compilerOptions: {
                ...options.compilerOptions,
                ...(_f = this.typescriptConfig) === null || _f === void 0 ? void 0 : _f.compilerOptions,
            },
        });
        this.config.preset = 'ts-jest';
        // only process .ts files
        this.config.testMatch = (_h = (_g = this.jestConfig) === null || _g === void 0 ? void 0 : _g.testMatch) !== null && _h !== void 0 ? _h : [
            '**/__tests__/**/*.ts?(x)',
            '**/?(*.)+(spec|test).ts?(x)',
        ];
        // specify tsconfig.json
        this.config.globals = {
            'ts-jest': {
                tsconfig: tsconfig.fileName,
            },
        };
        // add relevant deps
        this.project.addDevDeps('@types/jest', 'ts-jest');
        return tsconfig;
    }
    configureTestCommand() {
        const jestOpts = ['--passWithNoTests', '--all'];
        // if the project has anti-tamper configured, it should be safe to always run tests
        // with --updateSnapshot because if we forget to commit a snapshot change the CI build will fail.
        if (this.project.antitamper) {
            jestOpts.push('--updateSnapshot');
        }
        this.project.testTask.exec(`jest ${jestOpts.join(' ')}`);
        this.project.addTask('test:watch', {
            description: 'Run jest in watch mode',
            category: tasks_1.TaskCategory.TEST,
            exec: 'jest --watch',
        });
        this.project.addTask('test:update', {
            description: 'Update jest snapshots',
            category: tasks_1.TaskCategory.TEST,
            exec: 'jest --updateSnapshot',
        });
    }
}
exports.Jest = Jest;
