// projen/lib/ignore-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IgnoreFile = void 0;
const file_1 = require("./file");
/**
 * @experimental
 */
class IgnoreFile extends file_1.FileBase {
    /**
     * @experimental
     */
    constructor(project, filePath) {
        super(project, filePath, { editGitignore: filePath !== '.gitignore' });
        this.excludes = new Array();
        this.includes = new Array();
    }
    /**
     * @experimental
     */
    exclude(...patterns) {
        this.excludes.push(...patterns);
    }
    /**
     * @experimental
     */
    include(...patterns) {
        this.includes.push(...patterns);
    }
    /**
     * (experimental) Implemented by derived classes and returns the contents of the file to emit.
     *
     * @experimental
     */
    synthesizeContent(resolver) {
        return resolver.resolve([
            `# ${file_1.FileBase.PROJEN_MARKER}`,
            ...this.excludes,
            // includes must follow excludes
            ...this.includes.map(x => `!${x}`),
        ]).join('\n');
    }
}
exports.IgnoreFile = IgnoreFile;
