// projen/lib/github/workflows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GithubWorkflow = void 0;
const YAML = require("yaml");
const file_1 = require("../file");
/**
 * @experimental
 */
class GithubWorkflow extends file_1.FileBase {
    /**
     * @experimental
     */
    constructor(github, name) {
        super(github.project, `.github/workflows/${name.toLocaleLowerCase()}.yml`);
        this.events = {};
        this.jobs = {};
        this.name = name;
    }
    /**
     * @experimental
     */
    on(events) {
        this.events = {
            ...this.events,
            ...events,
        };
    }
    /**
     * @experimental
     */
    addJobs(jobs) {
        this.jobs = {
            ...this.jobs,
            ...jobs,
        };
    }
    /**
     * (experimental) Implemented by derived classes and returns the contents of the file to emit.
     *
     * @experimental
     */
    synthesizeContent(resolver) {
        const workflow = resolver.resolve({
            name: this.name,
            on: this.events,
            jobs: this.jobs,
        });
        return [
            `# ${file_1.FileBase.PROJEN_MARKER}`,
            YAML.stringify(workflow),
        ].join('\n');
    }
}
exports.GithubWorkflow = GithubWorkflow;
